### ISSUE TITLE
Limit camera view within map boundaries

### ISSUE THREAD

**Issue by MirceaKitsune** [Tuxemon/Tuxemon#373]
One of biggest visual problems still pertaining to the engine, is the lack of a cropping function to limit the camera view within the map boundaries as the player approaches an edge. This allows the player to see the nothingness beyond the world's limits, represented by a black strip that looks really ugly.

I created a visual example: The top image is an unedited screenshot. The bottom image was edited to represent how my suggestion would look in comparison, with the player standing on the same tile.

![screenshot](https://user-images.githubusercontent.com/825418/32415286-04810e3a-c240-11e7-8c15-1c7e813ec5ee.png)

This should be easy to fix with a simple bound check under the camera code:

```
camera_tile_x = min(camera_tile_x, map_tiles_x - (window_resolution_x / tile_size) / 2)
camera_tile_x = max(camera_tile_x, 0 + (window_resolution_x / tile_size) / 2)
camera_tile_y = min(camera_tile_y, map_tiles_y - (window_resolution_y / tile_size) / 2)
camera_tile_y = max(camera_tile_y, 0 + (window_resolution_y / tile_size) / 2)
```

**Comment by MirceaKitsune** [comment]
I feel slightly embarrassed for not noticing this before creating the issue, which I did as I was going to attempt a pull request for it; Pyscroll already offers a solution to this exact problem, which is simply disabled by a boolean. It's located at core/components/map.py line 325:

`return pyscroll.BufferedRenderer(visual_data, prepare.SCREEN_SIZE, clamp_camera=False)`

https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/core/components/map.py#L325

Is there any reason why we can't set clamp_camera to True? I gave this a very brief test, and it appears to work perfectly at a first glance.

**Comment by MirceaKitsune** [comment]
I feel slightly embarrassed for not noticing this before creating the issue, which I did as I was going to attempt a pull request for it; Pyscroll already offers a solution to this exact problem, which is simply disabled by a boolean. It's located at core/components/map.py line 325:

return pyscroll.BufferedRenderer(visual_data, prepare.SCREEN_SIZE, clamp_camera=False)

https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/core/components/map.py#L325

Is there any reason why we can't set clamp_camera to True? I gave this a very brief test, and it appears to work perfectly at a first glance.

@ShadowApex @bitcraft

**Comment by bitcraft** [comment]
Clamping the camera prevents the small maps like the bedroom from keeping
the player centered.  The map will stay stationary while the player sprite
moves.  That is one issue.  Another issue is that it allows the player
sprite outside of the center.  It might look odd, if the player expects the
sprite to stay centered.  The original Pokemon games had a camera similar
to the unclamped camera, where there are black edges when the map ends.
One fix would be to make a specific number of tiles on the border of
overworld maps to prevent the black areas.

**Comment by MirceaKitsune** [comment]
@bitcraft I don't see any of these as issues. If the map is smaller than the window resolution, it makes sense for the camera not to move with the player... which itself doesn't feel like an inconvenience.

This sounds more like a matter of preference then; Some players find the black borders better than the player being out of center, while others (myself included) find the black edges worse and have no problem with the player not always being in the middle as a compromise.

I'm therefore inclined to create a setting for this, which players can set in the config based on which model they like most. If there are no objections to that solution, I'll make a pull request for it later today. Do you consider this setting useful enough to be warranted, and should I default it to True at least?

**Comment by bitcraft** [comment]
I personally don't see a problem with a config option for the camera
clamp.  it won't prevent black edges on small maps still.  The default
should be left as-is, since it is more like the original Pokemon camera,
IMO, and fixing it on the overland maps is just a matter of adding more
content at the edges.  Maybe there are more opinions on it?

**Comment by Qiangong2** [comment]
This would make sense for regular overworld maps, but it wouldn't be needed for the insides of buildings, as the player already knows that there are walls and nothing on the other side except for the outside.

**Comment by MirceaKitsune** [comment]
Fixed in PR #374 I added a setting for camera clamping, defaulted to 0 as suggested.

@Qiangong2 Perhaps a map parameter to override this setting can also be added later. I only find the borders ugly and bothersome outdoors, but in buildings they feel perfectly natural.

**Comment by ShadowApex** [comment]
I think this sort of option should not be configurable from the configuration file, as it could allow a player to see the map differently from what a map creator might intend. For example, a map creator might be assuming part of a map is not visible, but with clamping enabled, would be visible. I think a better option would be to have a map action that could enable, disable, or toggle camera clamping. This would give map creators better control over the camera and would ensure that the experience is consistent across players.

In regards to the black strips on some maps, as @bitcraft mentioned, the maps themselves just need to be expanded to include extra tiles. Another alternative is implementing automatic map stitching in the engine, which would take a decent effort to implement.

**Comment by MirceaKitsune** [comment]
@ShadowApex If a map setting can be implemented instead of a global one, I definitely have no problem with that and quite prefer it! Like I said this should be used only on outdoor maps, but the engine needs to know what counts as such.

I took a look quickly, but unfortunately I don't know how to approach this implementation; It looks like a new entity for map properties might need to be implemented first, as there is currently no way of specifying map properties to my knowledge.

I don't think adding extra tiles in unreachable areas is a reliable alternative for a few reasons:

1. You need to surround every exterior map with impassible objects, in order to cover it in tiles the player can see but can't get to. This implies both having the right tiles, as well as having to design the architecture of a landscape differently just to avoid a visual issue.
2. How would you make areas that connect two maps, without having to use a door or a cave entrance? You can duplicate the edge of the neighboring map as an inaccessible visual of the current map, but that feels very messy and not something I think mappers should be doing.
3. Granted the window will ever be resizable and Tuxemon won't always use a fixed resolution, this is not a consisted fix as it depends on how big you expect players to make the window. For instance: Someone might design a map on a computer where the maximum screen resolution is 1600 x 1200, then test it with the window maximized and think "this looks good, no need to make my map bigger with extra tiles that no one should ever see"... but then someone with an 1920 x 1080 screen plays it and sees the annoying black borders.

**Comment by ShadowApex** [comment]
> I took a look quickly, but unfortunately I don't know how to approach this implementation; It looks like a new entity for map properties might need to be implemented first, as there is currently no way of specifying map properties to my knowledge.

Tiled has the ability to let you specify map properties:
![](https://i.imgur.com/UVhCjpD.png)

![](https://i.imgur.com/Wez1n6X.png)

In [core/components/map.py](https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/core/components/map.py#L212) in the `load` function is where the map data is actually loaded. The `self.data.properties` in that class contains the map properties that you specify in Tiled. From there, you can set the clampy-ness.

> You need to surround every exterior map with impassible objects, in order to cover it in tiles the player can see but can't get to. This implies both having the right tiles, as well as having to design the architecture of a landscape differently just to avoid a visual issue.

Maps don't need to be surrounded with impassible objects. It just needs to be either impassable or `teleport` actions that will load the next map.

> How would you make areas that connect two maps, without having to use a door or a cave entrance? You can duplicate the edge of the neighboring map as an inaccessible visual of the current map, but that feels very messy and not something I think mappers should be doing.

Same as above, you can just draw an area object in Tiled that will teleport you to another map when you reach close enough to the edge. I completely agree with you though that it should be easier for map creators to stitch together maps.

> Granted the window will ever be resizable and Tuxemon won't always use a fixed resolution, this is not a consisted fix as it depends on how big you expect players to make the window. For instance: Someone might design a map on a computer where the maximum screen resolution is 1600 x 1200, then test it with the window maximized and think "this looks good, no need to make my map bigger with extra tiles that no one should ever see"... but then someone with an 1920 x 1080 screen plays it and sees the annoying black borders.

Tuxemon scales up to whatever resolution the window is at, so there is a consistent view across resolutions.

**Comment by bitcraft** [comment]
The margins around the map is pretty simple.  You just need to have a boundary and some determined amount of space until the edge of the map.  The map1.tmx has this along the left/west edge...you never see the black areas because there is enough space.  As far as map stitching goes, there are a few comments I have about it.  my opinions, only.

- the rendering engine can handle arbitrarily large maps.  100 tiles or 10,000 will be the same, performance wise.  memory use is not a huge deal either.  instead of worrying about stitching maps, just combine all the overland maps into a large one.

- consider the legend of zelda a link to the past.  the overworld map is largely just one map, save a couple areas.  if map creators want to also combine maps into a large one, to reduce stitching burdens, then that is possible.

- while rendering maps may not suffer, the event engine is not tested with a large amount of events, as you would have in a single large overworld map.  in my brief tests i would say it is noting to worry about, however.  just better to be aware.

- automatic stitching is likely to be a fairly big process, as there are a lot of things to take into consideration, such as the map rendering, cleaning up of events, npcs, etc.  i don't see that feature being implemented for a long time.  my general goals for the project are to get the main features stable and implemented, then work on details, like stitching, as it is largely cosmetic.

- to map creators, i would advise not to spend time on stitching manually.  it is possible that it will be automatic in the future and the effort will be wasted.  maps will change often in development and stitching will just be busy work when real work could be done on the visible portions.

- if stitching needs to be done manually, it should be done only after the game is finished and ready for a non-beta, completely playable release.

- also, consider combining maps since there really is no limit to map size (in practical terms.  the real limit would be the size of python lists, which varies but is at least 536,870,912 x 536,870,912....larger on 64-bit systems...and probably larger than what tiled can handle)

@Sanglorian @Qiangong2

**Comment by ShadowApex** [comment]
@bitcraft My only concern with having a single large map is that in Tiled, you would have to deal with hundreds of events in a single map, which I think might be cumbersome for map creators. It also might be harder to collaborate with multiple map creators if they are editing a single large TMX map.

**Comment by bitcraft** [comment]
Good points @ShadowApex

**Comment by MirceaKitsune** [comment]
@ShadowApex Thanks for pointing this out! I didn't get to work with Tiled a lot yet, and since I didn't see any map properties yet I thought Tuxemon didn't support this. I'll take a look at it tomorrow hopefully, and will update my PR or make a new one then.

@bitcraft I was under the incorrect assumption that resizing the window would keep the resolution intact, thus changing the number of tiles you see on the screen. I realize now that would be weird, especially since there's no way to also re-scale the battle menu.

A single overworld map would be interesting if the engine also supported looped maps, which I assume is what you meant by stitching. I'm referring to a feature some RPG engines have (eg: RPG Maker, OHRRPGCE), where you can keep walking indefinitely in one direction and you'll always come out the other end of the map (simulates walking across the surface of a planet). I'd definitely love to see this at some point! But yeah, it will take a bit of work in PyScroll I imagine.

**Comment by bitcraft** [comment]
pyscroll just draws tiles and surfaces based on data from a "data adapter" (bad name, i know).  it would be up to the underlying data to represent looped maps or automatic stitching.  so, automatic stitching or looping is more about the underlying data, or tuxemon's responsibility.  in case you are interested, the link below is how pyscroll accesses tile data:

https://github.com/bitcraft/pyscroll/blob/master/pyscroll/data.py

**Comment by MirceaKitsune** [comment]
New PR #376 is up. It adds a simple framework for loading map properties, and of course a property to enable camera clamping on that map. The value can be set to "clamped" for this functionality, but is designed to support "stitched" for later when map stitching will be supported. I enabled it only on outdoor maps by default, since that feels a lot better and shouldn't harm anything if decorative border tiles are added later... if there's a strong opinion against this feel free to exclude the second commit of the merge.

**Comment by MirceaKitsune** [comment]
#376 is now merged, this issue has been resolved.
